common/lcs/src/trace.rs

```diff
@@ -0,0 +1,693 @@
+    Struct(Vec<Named<Format>>),
+}
+
+#[derive(Debug)]
+pub struct Tracer {
+    registry: BTreeMap<&'static str, Format>,
+}
+
+impl Tracer {
+    pub fn new() -> Self {
+        Self {
+            registry: BTreeMap::new(),
+        }
+    }
+
+    pub fn registry(&self) -> &BTreeMap<&'static str, Format> {
+        &self.registry
+    }
+
+    pub fn trace<T>(&mut self, value: &T) -> Result<Format>
+    where
+        T: ?Sized + Serialize,
+    {
+        value.serialize(self)
+    }
+}
+
+trait Merge {
+    /// Update `self` to replace unknown types or include newly discovered variants.
+    fn merge(&mut self, other: Self) -> Result<()>;
+}
+
+fn merge_error<T>(v1: T, v2: T) -> Error
+where
+    T: std::fmt::Debug,
+{
+    Error::Custom(format!("Incompatible formats detected: {:?} {:?}", v1, v2))
+}
+
+impl Merge for VariantFormat {
+    fn merge(&mut self, mut format: VariantFormat) -> Result<()> {
+        // Matching `&mut format` instead of `format` because of
+        // "error[E0009]: cannot bind by-move and by-ref in the same pattern"
+        // We can still make it work using std::mem::take.
+        match (&mut *self, &mut format) {
+            (Self::Unit, Self::Unit) => (),
+
+            (Self::NewType(format1), Self::NewType(format2)) => {
+                let format2 = std::mem::take(format2.as_mut());
+                format1.as_mut().merge(format2)?;
+            }
+
+            (Self::Tuple(formats1), Self::Tuple(formats2)) => {
+                if formats1.len() != formats2.len() {
+                    return Err(merge_error(formats1, formats2));
+                }
+                let mut formats2 = formats2.iter_mut();
+                for format1 in formats1 {
+                    let format2 = std::mem::take(formats2.next().unwrap());
+                    format1.merge(format2)?;
+                }
+            }
+
+            (Self::Struct(formats1), Self::Struct(formats2)) => {
+                if formats1.len() != formats2.len() {
+                    return Err(merge_error(formats1, formats2));
+                }
+                let mut formats2 = formats2.iter_mut();
+                for format1 in formats1 {
+                    let format2 = std::mem::take(formats2.next().unwrap());
+                    format1.merge(format2)?;
+                }
+            }
+
+            _ => {
+                return Err(merge_error(self, &mut format));
+            }
+        }
+        Ok(())
+    }
+}
+
+impl<T> Merge for Named<T>
+where
+    T: Merge + std::fmt::Debug,
+{
+    fn merge(&mut self, other: Named<T>) -> Result<()> {
+        if self.name != other.name {
+            return Err(merge_error(&*self, &other));
+        }
+        self.value.merge(other.value)
+    }
+}
+
+impl Merge for Format {
+    fn merge(&mut self, mut format: Format) -> Result<()> {
+        // Matching `&mut format` instead of `format` because of
+        // "error[E0009]: cannot bind by-move and by-ref in the same pattern"
```

Comment:
You should include a link to https://github.com/rust-lang/rust/issues/68354 which tracks the stabilization of this feature, move_ref_pattern.